Add MVar

alexandru opened the pull request:
Adds [MVar](https://hackage.haskell.org/package/base-4.9.1.0/docs/Control-Concurrent-MVar.html) to Cats-Effect.

PR also contains website documentation.

Implementation and documentation are imported from Monix, with some changes for working with `F[_]` and for cancelability.

SystemFw commented:
Only had a very quick glance, but one thing: you need `read` to be a primitive operation.
Having `read` be `take then put` leads to all sort of weird behaviour (mostly deadlock) due to concurrent interleaving (e.g. someone `puts` in between your `take` and your `put`).

The original MVar in Haskell had this issue, and `read` was changed to be primitive because of that.

The other question is adding derived operations, specifically which ones should be added (e.g. `modify`, `modifyF` or none) and related questions/docs about their nature (again, mostly due to them not being atomic).

alexandru commented:
@SystemFw 

> Only had a very quick glance, but one thing: you need read to be a primitive operation.
> Having read be take then put leads to all sort of weird behaviour (mostly deadlock) due to concurrent interleaving (e.g. someone puts in between your take and your put).

This is why I left `read` out of it. If you're designing `MVar` like a concurrent queue, which it is, `read` is difficult to introduce as an atomic operation.

Are there any `MVar` implementations that provide an atomic `read`? And is it worth the complexity cost?

> The other question is adding derived operations, specifically which ones should be added (e.g. modify, modifyF or none) and related questions/docs about their nature (again, mostly due to them not being atomic).

I would lean towards not including any derived operations.

SystemFw commented:
> Are there any MVar implementations that provide an atomic read? And is it worth the complexity cost?

Well, _the_ MVar implementation (the one in GHC) does, and it was deemed to be worth the cost due to the many problems implementing things that require reading.

> If you're designing MVar like a concurrent queue, which it is, read is difficult to introduce as an atomic operation.

I think this is an implementation detail that shouldn't influence the interface one way or another.

> I would lean towards not including any derived operations.

No strong opinions here, I can see pros and drawbacks of both

alexandru commented:
@SystemFw OK, will try to add it

alexandru commented:
@SystemFw so I did not remember `read` from `MVar` being atomic, but apparently this changed in "base 4.7": https://hackage.haskell.org/package/base-4.9.1.0/docs/Control-Concurrent-MVar.html#v:readMVar

SystemFw commented:
> so I did not remember read from MVar being atomic, but apparently this changed in "base 4.7"

Yeah that's what I meant, it proved too inconvenient/error prone in practice, and it was changed 👍

alexandru commented:
@SystemFw pushed a commit with an implementation that supports the atomic `read`.

I have to add more tests and maybe clean it up a bit.

SystemFw commented on core/shared/src/main/scala/cats/effect/concurrent/MVar.scala:
```diff
@@ -0,0 +1,162 @@
+ *  - [[take]] which empties the var if full, returning the contained
+ *    value, or blocks (asynchronously) otherwise until there is
+ *    a value to pull
+ *
```
`read` needs to be added here

alexandru commented:
I've added tests and documentation. It is now ready for review.

johnynek commented on core/shared/src/main/scala/cats/effect/internals/Callback.scala:
```diff
@@ -41,10 +41,10 @@ private[effect] object Callback {
     }
 
   /** Reusable `Right(())` reference. */
-  final val rightUnit = Right(())
```
why are these not final now?

alexandru replied:
AFAIK `final val` in companion objects will make the compiler treat that as a constant and inline it. Which would defeat the point of declaring it in the first place, as we're doing such shenanigans in order to not allocate memory every time we need a `Right(())`.

johnynek replied:
@non I thought told me that was only if there is no type annotation, so we could add that but keep the final. But in any case probably not a big deal.

non replied:
Yes, `final val x = ...` will inline but `final val x: SomeType = ...` will not. It's a weird rule.

alexandru replied:
Is there any difference between `final val x: SomeType = ...` and `val x: SomeType`?

From the JVM's point of view both will be `final`, so I'd rather go with the later.

johnynek commented on core/shared/src/main/scala/cats/effect/internals/LinkedMap.scala:
```diff
@@ -51,10 +55,20 @@ private[effect] class LinkedMap[K, +V](
+    emptyRef.asInstanceOf[LinkedMap[K, V]]
+
+  private val emptyRef =
+    new LinkedMap[Any, Any](Map.empty, LongMap.empty, 0)
```
strictly speaking I think `Nothing` would be a better type for the value.

johnynek commented on core/shared/src/main/scala/cats/effect/internals/MVarAsync.scala:
```diff
@@ -0,0 +1,197 @@
+  private def unsafePut(a: A)(onPut: Listener[Unit]): Unit = {
+    if (a == null) {
+      onPut(Left(new NullPointerException("null not supported in MVarAsync/MVar")))
+      return
```
rather than using `return` could we just wrap the rest in an `else`?

Reviewers have to always think "is this a return that is actually ok or not?" if we don't

johnynek commented on core/shared/src/main/scala/cats/effect/internals/MVarAsync.scala:
```diff
@@ -0,0 +1,197 @@
+      case current @ WaitForTake(value, puts) =>
+        val update = WaitForTake(value, puts.enqueue(a -> onPut))
+        if (!stateRef.compareAndSet(current, update)) {
+          // $COVERAGE-OFF$
```
do we need to do this? can we not try to trigger races that will actually hit this branch?

alexandru replied:
Triggering race conditions that hit those branches is very difficult and you're not gaining much.

johnynek replied:
It makes me nervous if we have concurrency primatives we aren’t testing in race conditions.

Seems like it shouldn’t be too hard to hit he branch: just have N threads and maybe two MVars that they take from one and put to the other or something like that.

oleg-py commented on core/shared/src/main/scala/cats/effect/concurrent/MVar.scala:
```diff
@@ -0,0 +1,165 @@
+      F.delay(MVarAsync[F, A](a)(F))
+
+    /** Builds an empty `MVar`. */
+    def empty[A](implicit F: Async[F]): F[MVar[F, A]] =
```
Why do you have a second `Async` here? You already have one as a class parameter.

alexandru replied:
Good catch. It's from my experiments 🙂

gvolpe commented on core/shared/src/main/scala/cats/effect/concurrent/MVar.scala:
```diff
@@ -0,0 +1,165 @@
+   *
+   * @see [[async]]
+   */
+  def apply[F[_]](implicit F: Concurrent[F]): ConcurrentBuilder[F] =
```
TBH I would prefer to have a simple `def apply[F[_]: Concurrent): MVar[F, A]`. I understand it brings some type inference help to `scalac` but on the other hand is less clearer to read and it is not consistent with the other datatypes recently introduced (`Ref`, `Deferred`, `Semaphore` and `Resource`).

gvolpe commented on site/src/main/tut/datatypes/mvar.md:
```diff
@@ -0,0 +1,156 @@
+  until there is a value available, at which point the operation resorts
+  to a `take` followed by a `put`
+
+An additional but non-atomic operation is `read`, which tries reading the
```
`read` is now an atomic operation? :)

alexandru commented:
@johnynek so regarding your review, I've added another concurrency test with `take`, `put` and `read` operations racing each other. I also removed the coverage annotations, as mentioned before.

@gvolpe 

> TBH I would prefer to have a simple def apply[F[_]: Concurrent): MVar[F, A]. I understand it brings some type inference help to scalac but on the other hand is less clearer to read and it is not consistent with the other datatypes recently introduced (Ref, Deferred, Semaphore and Resource).

In my experience thus far with Cats's monad transformers and with Monix's Iterant, without partially applying types in `apply`, these data types parameterized for `F[_]` become really annoying. As for the others:

- this isn't new and has been done in Cats, `EitherT` as mentioned for example
- `Ref` is addressed in #211 
- `Deferred` doesn't need it because you aren't giving it an initial value, so there's no type inference to be had for it
- `Semaphore` doesn't need it, because it doesn't have an `A` type parameter
- `Resource` doesn't need it because `apply` takes an `F[_]` as parameter and so the Scala compiler is perfectly fine inferring the types for that

Besides `Ref`, none of the other data types mentioned are relevant.

I did make changes. So in addition to `apply`'s magic, I added those builders as plain functions as well. So as mentioned in the ScalaDoc:

```scala
MVar[IO].empty[Int] <-> MVar.empty[IO, Int]

MVar[IO].init("hello") <-> MVar.init[IO, String]("hello")

MVar[IO].initF(IO("hello")) <-> MVar.initF(IO("Hello"))
```

mpilquist commented on core/shared/src/main/scala/cats/effect/concurrent/MVar.scala:
```diff
@@ -0,0 +1,216 @@
+   *
+   * @see [[init]], [[initF]] and [[empty]]
+   */
+  def apply[F[_]](implicit F: Concurrent[F]): ApplyBuilders[F] =
```
I like this pattern - can we use it in `Ref` too? Specifically the notion of having both the regular method on companion and then `ApplyBuilders` for the type curried type params.
